presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java

```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    @Override
+    public boolean isFull()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isBusy()
+    {
+        return !spillInProgress.isDone();
+    }
+
+    private boolean shouldSpill(long memorySize)
+    {
+        return (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+    }
+
+    private boolean shouldMergeWithMemory(long memorySize)
+    {
+        return memorySize < memoryLimitForMergeWithMemory;
+    }
+
+    public Iterator<Page> buildResult()
+    {
+        checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+        if (!spiller.isPresent()) {
+            return hashAggregationBuilder.buildResult();
+        }
+
+        try {
+            if (shouldMergeWithMemory(hashAggregationBuilder.getSizeInMemory())) {
+                return mergeFromDiskAndMemory();
+            }
+            else {
+                spillToDisk().get();
+                return mergeFromDisk();
+            }
+        }
+        catch (InterruptedException | ExecutionException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        if (merger.isPresent()) {
+            merger.get().close();
+        }
+        if (spiller.isPresent()) {
+            spiller.get().close();
+        }
+    }
+
+    private CompletableFuture<?> spillToDisk()
+    {
+        hashAggregationBuilder.setOutputPartial();
+
+        if (!spiller.isPresent()) {
+            spiller = Optional.of(spillerFactory.create(hashAggregationBuilder.buildTypes()));
+        }
+        long spillMemoryUsage = hashAggregationBuilder.getSizeInMemory();
+
+        // start spilling process with current content of the hashAggregationBuilder builder...
+        spillInProgress = spiller.get().spill(hashAggregationBuilder.buildResultSorted());
+        // ... and immediately create new hashAggregationBuilder so effectively memory ownership
+        // over hashAggregationBuilder is transferred from this thread to a spilling thread
+        rebuildHashAggregationBuilder();
+
+        // First decrease memory usage of aggregation context...
+        aggregationMemoryContext.setBytes(hashAggregationBuilder.getSizeInMemory());
+        // And then transfer this memory to spill context
+        // TODO: is there an easy way to do this atomically?
+        spillMemoryContext.setBytes(spillMemoryUsage);
+
+        return spillInProgress;
+    }
+
+    private Iterator<Page> mergeFromDiskAndMemory()
+    {
+        checkState(spiller.isPresent());
+
+        hashAggregationBuilder.setOutputPartial();
+
+        Iterator<Page> mergedSpilledPages = MergeSort.merge(
+                groupByTypes,
+                hashAggregationBuilder.buildIntermediateTypes(),
+                ImmutableList.<Iterator<Page>>builder()
+                        .addAll(spiller.get().getSpills())
```

Comment:
calling `getSpills()` will potentially allocate a lot of memory, since it has to allocate a `BufferedInputStream` for every spill. We're going to need to track that and limit it to avoid OOM'ing the workers. Is that something you're planning for a future PR, and if so can you file a issue and add a TODO in the code?